### DISCUSSION TITLE
How can I add multiple Site URLs?

### DISCUSSION THREAD

**Discussion by Saunved** [supabase/supabase#1514] [Questions] [answered]
Under Authentication > Settings, there is an option to add a site URL. **If I am doing local testing, I would want to put localhost:3000, but for development or production purpose, it should be example.com.**

I tried adding a comma-separated list but it broke the redirect post-registration.
Is there a simpler way to do this? I am sure this is a common use case.

**Edit**
Just saw this: https://github.com/supabase/supabase/discussions/600
This asks to create two projects (for dev and prod). 
I believe it would be much, much easier if a single project had a dev and prod mode. Or for now, if it simply allowed us to add multiple URLs, that would be beautiful.

**Comment by gomflo** [comment]
Also looking for this, for now i think the only way is to clone the project and have one project for dev and one for prod.

**Reply by Saunved** [reply]
@gomflo Thanks! How can I clone a project? I tried looking for this option, but didn't spot it easily anywhere.

**Reply by gomflo** [reply]
sorry theres no option to clone the project, what i mean is to create.a new project, then you can import the database thats what i mean with "clone".

**Reply by tomimarkus991** [reply]
Is this still the best way to do it?

**Reply by handlerda** [reply]
I am really hoping that this gets figured out. This hinders my workflow.

**Reply by geordidearns** [reply]
+1

**Reply by aquaductape** [reply]
I really need this, because I need ngrok workflow to be added

**Reply by libovness** [reply]
Is this is still the case? if I have `https://abc.xyz` as `Site URL`, can I also redirect to `https://def.xyz.com` if listed in `Additional redirect URLs`? Does not appear to work at this time

**Reply by mike623** [reply]
cannot use the site URL as comma list like `https://abc.com,http://localhost:3000/`

**Reply by lexome-neal** [reply]
If this feature can't be added easily, the copy on the dashboard should at least be changed.  The copy says the field is an allow "list", which implies multiple urls can be added.
<img width="562" alt="Screen Shot 2022-10-23 at 1 20 18 PM" src="https://user-images.githubusercontent.com/116481277/197416395-6591f9bf-7345-4e4f-a6ff-fee3f0b786a8.png">

**Comment by ctwhome** [comment]
I thought that this new addition would solve the problem, but it doen'st work.
![image](https://user-images.githubusercontent.com/4195550/118316520-37bd3280-b4f7-11eb-9ad7-9bc6422fbae6.png)

But it seems than not right? At least I didn't manage to handle it. This how my GCP OAuth looks like:
![image](https://user-images.githubusercontent.com/4195550/118316756-7f43be80-b4f7-11eb-8413-62a61eea88b3.png)

**Reply by ctwhome** [reply]
Firebase handles to redirect to localhost if the call comes from it, and from the production version otherwise with the same OAuth

**Comment by tomimarkus991** [comment]
Still waiting

**Comment by WesleyYue** [comment]
Adding multiple site URLs for magic links still won't work, but if you have just one site url, you can use the one-time passcode as a bypass and it will work despite the non-matching URL.

**Comment by alfredlua** [comment]
According to [this guide](https://supabase.com/docs/guides/auth#redirect-urls-and-wildcards), it seems like I could set my SITE_URL to my official URL and add localhost (http://localhost:3000/*/*) as an additional redirect URL in my Supabase Studio.

>When using third-party providers, the [Supabase client library](https://supabase.com/docs/reference/javascript/auth-signinwithoauth#sign-in-using-a-third-party-provider-with-redirect) redirects the user to the provider. When the third-party provider successfully authenticates the user, the provider redirects the user to the Supabase Auth callback URL where they are further redirected to the URL specified in the redirectTo parameter. This parameter defaults to the [SITE_URL](https://supabase.com/docs/reference/auth/config#site_url). You can modify the SITE_URL or [add additional redirect URLs](https://app.supabase.com/project/_/auth/settings).
>...
>For deployments with Vercel, set the SITE_URL to your official site URL. Add the following additional redirect URLs for local development and deployment previews:
>...

I also set the redirect in my signInWithGoogle function to my localhost:

```
async function signInWithGoogle() {
    const { data, error } = await supabase.auth.signInWithOAuth({
      provider: 'google',
      options: {
        redirectTo: 'http://localhost:3000/'
      }
    })
  }
```

However, when I sign in with Google to my app locally, I get redirected to my official URL post authentication instead of my localhost. 

Did anyone figure out a solution, besides having two projects? Thanks!

**Reply by thorwebdev** [reply]
Can you try setting `http://localhost:3000` as an additional redirect URL, and specify the `redirectTo` without the trailing `/`?

I think unfortunately the matcher does not match on `''`

**Reply by thorwebdev** [reply]
Looking at https://www.digitalocean.com/community/tools/glob?comments=true&glob=http%3A%2F%2Flocalhost%3A3000%2F%2A%2A&matches=false&tests=http%3A%2F%2Flocalhost%3A3000&tests=http%3A%2F%2Flocalhost%3A3000%2F%3Ftest%3Dtest&tests=http%3A%2F%2Flocalhost%3A3000%2Ftest-test%3Ftest%3Dtest&tests=http%3A%2F%2Flocalhost%3A3000%2Ftest%2Ftest%3Ftest%3Dtest then `http://localhost:3000/**` might be the best pattern to recommend to most folks. I'll update the docs.

**Reply by alfredlua** [reply]
1. Setting `http://localhost:3000` as an additional redirect URL and specifying the `redirectTo` without the trailing `/` works!
2. Setting `http://localhost:3000/**` as a redirect URL and removing `http://localhost:3000` works too.

Thanks for the quick help, @thorwebdev!

**Reply by Ve2s4** [reply]
Using a glob matcher URL for localhost (e.g. http://localhost:3000/**) in the redirect URI and production URL in the site URL worked for me. 

If you are following this [guide](https://github.com/vercel/nextjs-subscription-payments) and trying to deploy on EC2 instance instead of Vercel, make sure to configure the following for the auth/callback url

```
import { createClient } from '@/utils/supabase/server';
import { NextResponse } from 'next/server';
import { NextRequest } from 'next/server';
import { getErrorRedirect } from '@/utils/helpers';

export async function GET(request: NextRequest) {
    // The `/auth/callback` route is required for the server-side auth flow implemented
    // by the `@supabase/ssr` package. It exchanges an auth code for the user's session.
    const requestUrl = new URL(request.url);
    const code = requestUrl.searchParams.get('code');

    if (code) {
        const supabase = createClient();

        const { error } = await supabase.auth.exchangeCodeForSession(code);

        if (error) {
            console.log('error while exchanging the code for token:', error);
            return NextResponse.redirect(
                getErrorRedirect(
                    `${requestUrl.origin}/signin`,
                    error.name,
                    "Sorry, we weren't able to log you in. Please try again.",
                ),
            );
        }
        console.log('redirecting to:', requestUrl.origin);
    }

    const cookieMap = Object.fromEntries(
        request.headers
            .get('cookie')
            ?.split('; ')
            .filter((row) => row.startsWith('teamId=') || row.startsWith('email='))
            .map((row) => row.split('=')) || [],
    );

    const isEmpty = (obj: { [key: string]: string }) => Object.keys(obj).length === 0;

    // URL to redirect to after sign in process completes
    return NextResponse.redirect(
        isEmpty(cookieMap)
            ? '${requestUrl.origin}'
            : `${requestUrl.origin}/invite?teamId=${cookieMap.teamId}&email=${cookieMap.email}`,
    );
}
```
Instead of requestUrl.origin use you production domain like `https://example.com`, because it resolves to the internal address (0.0.0.0:3000) when running on EC2.

**Comment by thorwebdev** [comment, marked as answer]
Documentation for this has been added here: https://supabase.com/docs/guides/auth/overview#redirect-urls-and-wildcards

**Reply by EskelCz** [reply]
The link is dead :-/

**Reply by thorwebdev** [reply]
Sorry, moved to https://supabase.com/docs/guides/auth/overview#redirect-urls-and-wildcards will add a redirect.

**Reply by rkristelijn** [reply]
It has moved again to https://supabase.com/docs/guides/auth/redirect-urls

**Comment by abhishekjakhar** [comment]
The redirect to localhost:3000 is working for me. 
Thank you very much!

**Comment by gclcodeguy** [comment]
How do you achieve the same thing when using Auth Ui?

**Reply by vikram-ray** [reply]
This works in Auth UI
```
<Auth
    supabaseClient={supabase}
    appearance={{ theme: ThemeSupa }}
    providers={["google"]}
    redirectTo="http://localhost:3000"
  />
```

**Comment by rkristelijn** [comment]
My use case is a bit more complex using NextJS and [Keycloak](https://supabase.com/docs/guides/auth/social-login/auth-keycloak) as authentication. I got it al working using Keycloak (cloud) and Supabase (cloud) with the following config (see [github](https://github.com/rkristelijn/supabase-keycloak-auth) ):
- Site URL: http://localhost:3000
- Redirect URLs: ['http://localhost:3000/callback']
- in code: redirectTo: set to http://localhost:3000/callback

It works great.

Now when I deployed this to Cloudflare in a preview environment, it redirects to localhost:3000. Silly me, forgot to update the allowed redirect list, so I added https://*.my-awesome-project.pages.dev/ (my preview env is https://24d07fc3.my-awesome-project.pages.dev) to the allowed list on Redirect URLs, tested it again, but it keeps on redirecting to localhost:3000 because of the Site URL, even though I ship the proper value in redirectTo in the call and the right value is in the Redirect URLs in Supabase. Once I change the Site URL to https://24d07fc3.my-awesome-project.pages.dev it works. ٩(ˊᗜˋ*)و 

It seems it is locked to the Site URL. Which is kind of a problem, because it doesn't allow wildcards, so everytime I spin up a new preview environment, I have to go in and change the Site URL. On top of that I can only use one preview environment, and what about local? I want that to work too. ε(´סּ︵סּ`)з 

I can imagine in need a different environment for localhost, one for production, but what about preview environments? I have many and it would be nice to not having to go in and change the Site URL to match my preview environment, breaking preview environments of co-workers on the go. (;´༎ຶД༎ຶ`) 

My question: why is Supabase using Site URL over my provided redirectTo? ¯\(°_o)/¯

**Reply by rkristelijn** [reply]
With help I got this working now, in Supabase Authentication / Url Configuration:

Site URL: `https://my-project.pages.dev`

Redirect URLs:
- `https://**my-project.pages.dev/**`
- `http://localhost:3000/**`

I validated proper workings on:
- `https://feature-supabase-keycloak-lo.my-project.pages.dev/`
- `https://0b051b6c.my-project.pages.dev/`
- `http://localhost:3000`

; after logging in via Keycloak, it redirects to Supabase and then to the url it came from.

**Reply by pablodz** [reply]
Also works with local nets, thanks @rkristelijn

**Reply by desoga10** [reply]
This is the perfect solution. Thanks @rkristelijn

**Reply by SeungJin051** [reply]
It's working well. Thank you. @rkristelijn

**Reply by 7TIN** [reply]
Thank you @rkristelijn for best solution.
